### PR TITLE
Synchronize parallel smallfile operations and samples using ripsaw redis sidecar

### LINKED ISSUE #98: smallfile pods not synchronized
I've been reminded (by using elastic search and kibana!) that smallfile ripsaw pods are not synchronizing with each other, so if you request a sequence of operations, such as create,read,append,delete and multiple samples of each, pod 1 may start reads before pod 2 has finished creates, etc. and over time this becomes worse and worse, preventing us from really measuring what throughput and response time is generated by each operation type.     The solution is probably to use redis to synchronize the different pods so that pod 1 doesn't start its read until pods other than pod 1 have finished their creates, and so on.  smallfile is still usable to some extent but it would be better and more scalable if this was fixed.

### PR DESCRIPTION
**Pull request by rsevilla87**
Fixes https://github.com/cloud-bulldozer/snafu/issues/98 together with https://github.com/cloud-bulldozer/ripsaw/pull/259
The code here uses the ripsaw's redis to synchronize smallfile samples and operations, so that all of them will be executed in parallel.

Closes #98

### CODE REVIEW

**Comment by aakarshg** [comment]
ok to test

**Comment by rsevilla87** [comment]
What failed here?, I've been able to run `./smallfile_wrapper/ci_test.sh` successfully in my local.

File: smallfile_wrapper/trigger_smallfile.py [hunk 1]
```diff
@@ -137,6 +141,23 @@ def emit_actions(self):
                             interval['99%'] = float(flds[10])
                             yield interval, '-rsptimes'
 
+            if self.clients > 1:
+                self.logger.info("Number of smallfile pods > 1, synchronizing them for the next operation")
```
Comment (jtaleric) [inline smallfile_wrapper/trigger_smallfile.py]: There is never a case we _don't_ want them to be synchronized?

File: smallfile_wrapper/trigger_smallfile.py [hunk 2]
```diff
@@ -137,6 +141,23 @@ def emit_actions(self):
...
+                r = redis.StrictRedis(self.redis_host, 6379)
+                p = r.pubsub()
+                p.subscribe("smallfile")
+                subscribers = int(r.pubsub_numsub("smallfile")[0][1])
+                if subscribers == self.clients:
```
Comment (jtaleric) [inline smallfile_wrapper/trigger_smallfile.py]: Do we want to add a limit on the number of times we check to see if the clients are synced? For example if a client pod goes into OOMKill or some other undesirable fault, should we exit notifying the user there is a problem?

File: smallfile_wrapper/trigger_smallfile.py (hunk 1 above)
Comment (rsevilla87) [reply smallfile_wrapper/trigger_smallfile.py]: In case we perform the benchmark with only one client, we skip sync code.

**Comment by jtaleric** [comment]
@aakarshg @bengland2 LGTM can someone else throw eyes on this?

File: smallfile_wrapper/trigger_smallfile.py (hunk 2 above)
Comment (rsevilla87) [reply smallfile_wrapper/trigger_smallfile.py]: I've added a redis timeout to the implementation which will make redis subscription to timeout if no message is received on that channel. By default, 60 seconds. Can be  overridden by an env. var

**Comment by rsevilla87** [comment]
This PR depends on https://github.com/cloud-bulldozer/ripsaw/pull/259
```
  File "/opt/snafu/smallfile_wrapper/smallfile_wrapper.py", line 59, in __init__
    self.redis_host = os.environ["redis_host"]
  File "/usr/lib64/python2.7/UserDict.py", line 23, in __getitem__
    raise KeyError(key)
```

Once the previous is merged, we can move forward with this.

**Comment by aakarshg** [comment]
needs a rebase and resolving conflicts.

**Comment by rsevilla87** [comment]
> needs a rebase and resolving conflicts.

Done ;-)

**Comment by aakarshg** [comment]
@bengland2 can you PTAL and give the final approval :)

**Comment by rsevilla87** [comment]
> sorry for the delay in the review. in the tests/test_crs/valid_smallfile.yaml CR , I see clients: 1. Therefore the code in this pull request will not get tested by the CI, right? Have you tested it separately? Should it be tested by the CI? If so, then let's change clients from 1 to 2.

Yes, i tested it separately. In case of adding a test here... what do you suggest?, Do we have to add an additional test or modify the current one?

**Comment by bengland2** [comment]
@rsevilla87 I'm suggesting a 1 line change to valid_smallfile.yaml so this code path is tested, that's all.

**Comment by rsevilla87** [comment]
Thanks Ben, that's another PR on ripsaw. Let's wait for have this one merged

**Comment by aakarshg** [comment]
> Thanks Ben, that's another PR on ripsaw. Let's wait for have this one merged

feel free to open a pr and tag me, i'll try to get it merged asap :)

**Comment by rsevilla87** [comment]
What's missing on this PR to get merged?

**Comment by bengland2** [comment]
BTW usually you create a branch for your PR instead of using master in your forked repo.   Especially since you have multiple active PRs.  But no big deal, I'll run it anyway.

**Comment by bengland2** [comment]
doesn't look like your fault @rsevilla87 that linters CI job failed, a different PR had exact same error.

**Comment by bengland2** [comment]
I'm confused, is this being thrown away or is it still active?  I thought we had sorted this out and it was ready for merge, it only affects smallfile so no big deal right?

**Comment by rsevilla87** [comment]
> I'm confused, is this being thrown away or is it still active? I thought we had sorted this out and it was ready for merge, it only affects smallfile so no big deal right?

My fault, I mistakenly closed it.

**Comment by rsevilla87** [comment]
I'm working in some minor changes to avoid snafu to hang when redis_host variable is not defined to make Snafu able to run even outside of ripsaw.  I'll make you know when ready.

**Comment by rsevilla87** [comment]
I had to change some details of the implementation. Since the migration to python3, redis channel messages are now bytes instead of strings.

**Comment by aakarshg** [comment]
@rsevilla87 can you please fix the linting issue and also rebase this ? I am not able to do any merge because of conflict/.

**Comment by bengland2** [comment]
@rsevilla87  ok, I'll take a look while I wait for my cluster to be repaired...

**Comment by bengland2** [comment]
@rsevilla87 this conflict is trivial to resolve, just adding the "self.redis_host, self.redis_timeout, self.clients, " arguments to the trigger_smallfile call + some reformatting.
```
[bengland@localhost snafu]$ git diff
diff --cc smallfile_wrapper/smallfile_wrapper.py
index e45e1be,66cf5f7..0000000
--- a/smallfile_wrapper/smallfile_wrapper.py
+++ b/smallfile_wrapper/smallfile_wrapper.py
...
@@@ -74,10 -70,10 +73,8 @@@
              sample_dir = self.result_dir + '/' + str(s)
              if not os.path.exists(sample_dir):
                  os.mkdir(sample_dir)
 -            trigger_generator = trigger_smallfile._trigger_smallfile(logger, self.operations,
 -                                                                     self.yaml_input_file,
 -                                                                     self.cluster_name,
 -                                                                     self.working_dir, sample_dir,
 -                                                                     self.user, self.uuid,
 -                                                                     s)
 +            trigger_generator = trigger_smallfile._trigger_smallfile( 
 +                    logger, self.operations, self.yaml_input_file,
 +                    self.cluster_name, self.working_dir, sample_dir, self.user,
 +                    self.uuid, self.redis_host, self.redis_timeout, self.clients, s)
              yield trigger_generator
-     
-  
```

File: smallfile_wrapper/trigger_smallfile.py [hunk 3]
```diff
@@ -5,6 +5,7 @@
 import shutil
 import socket
 from vfs_stat import get_vfs_stat_dict
+import redis
```
Comment (aakarshg) [inline smallfile_wrapper/trigger_smallfile.py]: Can you please add this library to the requirements as well ? https://github.com/cloud-bulldozer/snafu/blob/master/requirements.txt

**Comment by rsevilla87** [comment]
@aakarshg added some doc & diagram. PTAL

File: uperf-wrapper/README.md [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# Smallfile
```
Comment (aakarshg) [inline uperf-wrapper/README.md]: I think you meant to add this to smallfile readme file but added it uperf by mistake ?

Comment (rsevilla87) [reply uperf-wrapper/README.md]: woops, you're right, nice catch :)

**Comment by aakarshg** [comment]
@rsevilla87 can you please rebase this ?

**Pull request merged**